velox/expression/CastExpr-inl.h

```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
       std::current_exception(),
       makeErrorMessage(input, row, resultType, errorDetails),
       false));
+}
+
+/// Represent the varchar fragment.
+///
+/// For example:
+/// | value | wholeDigits | fractionalDigits | exponent | sign
+/// | 9999999999.99 | 9999999999 | 99 | nullopt | 1
+/// | 15 | 15 |  | nullopt | 1
+/// | 1.5 | 1 | 5 | nullopt | 1
+/// | -1.5 | 1 | 5 | nullopt | -1
+/// | 31.523e-2 | 31 | 523 | -2 | 1
+struct DecimalComponents {
+  std::string_view wholeDigits;
+  std::string_view fractionalDigits;
+  std::optional<int32_t> exponent = std::nullopt;
+  int8_t sign = 1;
 };
 
+size_t parseDigitsRun(
+    const char* s,
+    size_t start,
+    size_t size,
+    std::string_view& out) {
+  size_t pos = start;
+  for (; pos < size; ++pos) {
+    if (!std::isdigit(s[pos])) {
+      break;
+    }
+  }
+  out = std::string_view(s + start, pos - start);
+  return pos;
+}
+
+std::optional<DecimalComponents> parseDecimalComponents(
+    const char* s,
+    size_t size) {
+  if (size == 0) {
+    return std::nullopt;
+  }
+  DecimalComponents out;
+  size_t pos = 0;
+  // Sign of the number.
+  if (s[pos] == '-') {
+    out.sign = -1;
+    ++pos;
+  } else if (s[pos] == '+') {
+    out.sign = 1;
+    ++pos;
+  }
+  // First run of digits.
+  pos = parseDigitsRun(s, pos, size, out.wholeDigits);
+  if (pos == size) {
+    return out.wholeDigits.empty() ? std::nullopt
+                                   : std::optional<DecimalComponents>(out);
+  }
+  // Optional dot (if given in fractional form).
+  if (s[pos] == '.') {
+    // Second run of digits.
+    ++pos;
+    pos = parseDigitsRun(s, pos, size, out.fractionalDigits);
+  }
+  if (out.wholeDigits.empty() && out.fractionalDigits.empty()) {
+    // Need at least some digits (whole or fractional).
+    return std::nullopt;
+  }
+  if (pos == size) {
+    return out;
+  }
+  // Optional exponent.
+  if (s[pos] == 'e' || s[pos] == 'E') {
+    ++pos;
+    if (pos != size && s[pos] == '+') {
```

Comment:
Is it not possible to have a '-' here?

Reply:
Here we convert `+10` to `10`, then convert it to int32 by folly, convert `-10` by folly too.